src/main/java/org/pac4j/vertx/handler/impl/RequiresAuthenticationHandler.java

```diff
@@ -39,73 +46,130 @@
-    } else {
-      final VertxWebContext webContext = new VertxWebContext(routingContext);
-      final Client client = findClient(webContext, this.clientName, allowDynamicClientSelection);
-      final ProfileManager profileManager = new ProfileManager(webContext);
-      final AuthenticationFlow authenticationFlow = getAuthenticationFlow(client);
-
-      UserProfile profile = profileManager.get(authenticationFlow.useSession(webContext));
-      if (profile != null) {
-        // We have been authenticated, are we authorized? Use vert.x subsystem for this
-        authorise(new Pac4jUser(profile), routingContext);
-      } else {
-        // We have not been authenticated yet, so start of the authentication process
-        authenticationFlow.initiate(webContext, new Consumer<UserProfile>() {
-          @Override
-          public void accept(UserProfile userProfile) {
-            final Pac4jUser pac4jUser = new Pac4jUser(userProfile);
-            if (userProfile != null) {
-              profileManager.save(authenticationFlow.useSession(webContext), userProfile);
-              // Writing into the routing context now means that if the UserSessionHandler is in use then nothing
-              // else needs to be done to retrieve the user from the session and write to the routing context
-              routingContext.setUser(pac4jUser);
+    private static final Logger LOG = LoggerFactory.getLogger(RequiresAuthenticationHandler.class);
+
+    protected final Config config;
+    protected final String clientName;
+    protected final Vertx vertx;
+
+    protected HttpActionAdapter httpActionAdapter = new DefaultHttpActionAdapter();
+    protected ClientFinder clientFinder = new DefaultClientFinder();
+
+    public RequiresAuthenticationHandler(final Vertx vertx, final Config config, final Pac4jAuthProvider authProvider,
+                                         final Pac4jAuthHandlerOptions options) {
+        super(authProvider);
+        clientName = options.clientName();
+        this.vertx = vertx;
+        this.config = config;
+    }
+
+    // Port of Pac4J auth to a handler in vert.x 3.
+    @Override
+    public void handle(RoutingContext routingContext) {
+
+        final User user = routingContext.user();
+        if (user != null) {
+            // Already logged in, just authorise
+            authorise(user, routingContext);
+        } else {
+            final VertxWebContext webContext = new VertxWebContext(routingContext);
+            final List<Client> currentClients = clientFinder.find(config.getClients(), webContext, this.clientName);
+            final ProfileManager profileManager = new ProfileManager(webContext);
+
+            UserProfile profile = profileManager.get(useSession(webContext, currentClients));
+            if (profile != null) {
+                // We have been authenticated, are we authorized? Use vert.x subsystem for this
+                authorise(new Pac4jUser(profile), routingContext);
```

Comment:
Yes, because I’m trying to use the vert.x infrastructure for this. If we don’t take this approach, we will actually break a piece of behaviour which will be familiar to vert.x coders (and is part of the standard vertx-provided infrastructure) - the UserSessionHandler. 

> inal User user = routingContext.user();
> > +        if (user != null) {
> > +            // Already logged in, just authorise
> > +            authorise(user, routingContext);
If we don’t include this, we make it impossible for a vertx coder to use the UserSessionHandler, which would break something familiar to vertx coders. I see that as breaking a contract with the platform.


> On 28 Sep 2015, at 11:16, LELEU Jérôme <notifications@github.com> wrote:
> 
> In src/main/java/org/pac4j/vertx/handler/impl/RequiresAuthenticationHandler.java <https://github.com/pac4j/vertx-pac4j/pull/45#discussion_r40536857>:
> 
> > +    @Override
> > +    public void handle(RoutingContext routingContext) {
> > +
> > +        final User user = routingContext.user();
> > +        if (user != null) {
> > +            // Already logged in, just authorise
> > +            authorise(user, routingContext);
> > +        } else {
> > +            final VertxWebContext webContext = new VertxWebContext(routingContext);
> > +            final List<Client> currentClients = clientFinder.find(config.getClients(), webContext, this.clientName);
> > +            final ProfileManager profileManager = new ProfileManager(webContext);
> > +
> > +            UserProfile profile = profileManager.get(useSession(webContext, currentClients));
> > +            if (profile != null) {
> > +                // We have been authenticated, are we authorized? Use vert.x subsystem for this
> > +                authorise(new Pac4jUser(profile), routingContext);
> I see you check authorizations in two places (here and at line 116). I tend to think that it makes source code a little harder to read. I prefer the J2E / Play algorithm, which works first on the user profile retrieval and then on access checks.
> 
> In Play, the user profile is returned as a promise wherever we get it. Based on the promised result, checks happen.
> 
> Maybe I'm nitpicking. Ignore this comment if you feel so.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/pac4j/vertx-pac4j/pull/45/files#r40536857>.
>